### ISSUE TITLE
Expose mesh partitioner in the C++ API

### ISSUE THREAD

**Issue by IgorBaratta** [FEniCS/dolfinx#1200]
The aim of this issue is to discuss the proper way to expose the choice of mesh partitioner.

Although dolfinx has support for three different graph partitioners, only SCOTCH is currently available to the user.
Previously a mesh partitioner `ENUM` was exposed from the high-level mesh generation interface (BoxMesh, Mesh ..) and passed all the way to a more low-level `mesh::partition_cells` function.

I particularly prefer a leaner interface, with the mesh partitioner `ENUM` exposed only as an argument to `mesh::partition_cells`.
But that would prevent using, for example, `UnitCubeMesh` directly with ParMETIS or KaHIP.  As another option we could add `mesh::BoxMesh::create_geom` and `mesh::BoxMesh::create_topology`, and use a low level interface:

```
  auto topo = mesh::BoxMesh::create_topology(args ...)
  ...
  graph::AdjacencyList<std::int64_t> cells_topology(topo);

  auto dest = dolfinx::mesh::Partitioning::partition_cells(
        MPI_COMM_SELF, nparts, element.cell_shape(),
        cells_topology, ghost_mode);

  const auto [cell_nodes, src, original_cell_index, ghost_owners]
      = graph::Partitioning::distribute(comm, cells_topology, dest, **partitioner**);

  dolfinx::mesh::Topology topology
      = mesh::create_topology(comm, cell_nodes, original_cell_index,
                              ghost_owners, element.cell_shape(), ghost_mode);

  dolfinx::mesh::Geometry geometry
      = mesh::create_geometry(comm, topology, element, cell_nodes, geom);

  dolfinx::mesh::Mesh(comm, std::move(topology), std::move(geometry));
``` 

@garth-wells , @chrisrichardson , @michalhabera 
Any thoughts on that?

**Comment by michalhabera** [comment]
Do we really expect anyone who is using a high-level mesh generation interface to switch the partitioner? I doubt it. It is certainly more important to have the choice in mesh IO.

**Comment by IgorBaratta** [comment]
The only use case I can see is in `performance-test`, since it uses `BoxMesh` to create the mesh.

**Comment by garth-wells** [comment]
I think we should start here from the lowest level, which is what is the common interface that all partitioners should have and should this be just a function signature or should it be a base class? The user should then be able to pass the partitioning function to the function that requires it. From this point we can build up the interface. I would anticipate enums coming last.

**Comment by chrisrichardson** [comment]
Maybe we could use a `std::function` then, since the partitioner does not have any state.

**Comment by IgorBaratta** [comment]
Fixed in #1287 .
